### PR TITLE
Rewrite build and upload scripts in Dart

### PR DESCRIPTION
**Pull request by bryanoltman**
A fairly literal translation of our [platform]_build_and_upload scripts to Dart. There are a number of things that can/should be cleaned up or refactored, some of which I have left TODOs for.

I've verified that the "build" portion of "build and upload" works on mac, but I have not tested the "upload" portion as I don't want to do anything accidentally destructive. I have, however, verified that the script produces the same set of build and upload commands on mac, modulo:

- Temp directory names
- Absolute vs relative paths
- Flag ordering for gn commands (I tried to standardize those a bit between commands)
- The `-j 4` flags on the ninja builds. We can reintroduce those if we'd like, but at least on my Mac Studio, it was slowing down the build unnecessarily.

### CODE REVIEW

File: bin/checkout.dart [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import 'package:build_engine/cli.dart';
+import 'package:cli_script/cli_script.dart';
+import 'package:mason_logger/mason_logger.dart';
+import 'package:path/path.dart' as p;
+
+final cli = Cli(logger: Logger());
+
+void main(List<String> arguments) {
+  final argParser = ArgParser();
+  // ignore: cascade_invocations
```
Comment (felangel) [inline bin/checkout.dart:13]: Why not cascade and remove this ignore?

Comment (bryanoltman) [reply bin/checkout.dart]: Fixed.

File: pubspec.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  sdk: ">=3.0.0 <4.0.0"
+
+dependencies:
+  args: ^2.4.1
+  cli_script: ^1.0.0
+  mason_logger: ^0.2.5
+  path: ^1.8.3
+
+dev_dependencies:
+  test: ^1.24.2
```
Comment (felangel) [inline pubspec.yaml:16]: Do we need this if there aren't any tests?

**Review by felangel — APPROVED** [review]
LGTM overall 👍

**Review by eseidel — COMMENTED** [review]
Oh, I somehow never sent my review comments, they're mostly silly.

File: pubspec.yaml (hunk 2 above)
Comment (bryanoltman) [reply pubspec.yaml]: Nope, but we'll obviously want some eventually

File: bin/build_and_upload.dart [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      cli.workingDirectory = paths.engineFlutterDirectory;
+      final baseEngineTag = await cli.output('git describe --tags --abbrev=0');
+      final baseEngineHash = await cli.output('git rev-parse $baseEngineTag');
+
+      if (Platform.isMacOS) {
+        await builder.buildUpdater();
+        await builder.buildPatchTool();
+        await builder.ninjaBuild(
+          architectureTargets: {
+            BuildArchitecture.androidArm32: [],
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply bin/build_and_upload.dart]: I documented this on the `ninjaBuild` function ("If no targets are specified, the default targets are built"), but lmk if I can make this clearer

File: bin/checkout.dart [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply bin/checkout.dart]: Yep, this has been fixed

File: bin/build_and_upload.dart [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  final engineRoot = Directory(parsedArgs[engineRootArg] as String);
+
+  final paths = Paths(engineRoot: engineRoot);
+  final cli = Cli(logger: Logger(level: verbose ? Level.verbose : Level.info));
+  final builder = Builder(
+    cli: cli,
+    paths: paths,
+  );
+  final uploader = Uploader(
+    cli: cli,
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply bin/build_and_upload.dart]: +1

File: bin/build_and_upload.dart [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    engineHash: engineHash,
+  );
+
+  wrapMain(
+    () async {
+      cli.workingDirectory = paths.engineFlutterDirectory;
+      final baseEngineTag = await cli.output('git describe --tags --abbrev=0');
+      final baseEngineHash = await cli.output('git rev-parse $baseEngineTag');
+
+      if (Platform.isMacOS) {
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply bin/build_and_upload.dart]: Done

File: lib/build_architecture.dart [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      };
+
+  /// Only valid for Android targets
+  Directory? get zipArchivesDirectory {
+    final zipDirName = switch (this) {
+      BuildArchitecture.androidArm32 => 'android-arm-release',
+      BuildArchitecture.androidArm64 => 'android-arm64-release',
+      BuildArchitecture.androidx64 => 'android-x64-release',
+      BuildArchitecture.androidx86 => 'android-x86-release',
+      BuildArchitecture.iosArm64 => null,
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply lib/build_architecture.dart]: One of many "am I using the right abstraction? am I totally misunderstanding how all of this works?" moments :)

**Pull request merged**